sdk/go/picovoice.go

```diff
@@ -0,0 +1,266 @@
+	RhinoVersion = rhn.Version
+
+	// Picovoice version
+	Version = fmt.Sprintf("1.1.0 (Porcupine v%s) (Rhino v%s)", PorcupineVersion, RhinoVersion)
+)
+
+// Init function for Picovoice. Must be called before attempting process.
+func (picovoice *Picovoice) Init() error {
+
+	if picovoice.KeywordPath == "" {
+		return fmt.Errorf("%s: No valid keyword was provided.", pvStatusToString(INVALID_ARGUMENT))
+	}
+
+	if _, err := os.Stat(picovoice.KeywordPath); os.IsNotExist(err) {
+		return fmt.Errorf("%s: Keyword file file could not be found at %s", pvStatusToString(INVALID_ARGUMENT), picovoice.KeywordPath)
+	}
+
+	if picovoice.ContextPath == "" {
+		return fmt.Errorf("%s: No valid context was provided.", pvStatusToString(INVALID_ARGUMENT))
+	}
+
+	if _, err := os.Stat(picovoice.ContextPath); os.IsNotExist(err) {
+		return fmt.Errorf("%s: Context file could not be found at %s", pvStatusToString(INVALID_ARGUMENT), picovoice.ContextPath)
+	}
+
+	if ppn.SampleRate != rhn.SampleRate {
+		return fmt.Errorf("%s: Pocupine sample rate (%d) was differenct than Rhino sample rate (%d)",
+			pvStatusToString(INVALID_ARGUMENT),
+			ppn.SampleRate,
+			rhn.SampleRate)
+	}
+
+	if ppn.FrameLength != rhn.FrameLength {
+		return fmt.Errorf("%s: Pocupine frame length (%d) was differenct than Rhino frame length (%d)",
+			pvStatusToString(INVALID_ARGUMENT),
+			ppn.FrameLength,
+			rhn.FrameLength)
+	}
+
+	picovoice.porcupine = ppn.Porcupine{
+		ModelPath:     picovoice.PorcupineModelPath,
+		KeywordPaths:  []string{picovoice.KeywordPath},
+		Sensitivities: []float32{0.5},
+	}
+	err := picovoice.porcupine.Init()
+	if err != nil {
+		return err
+	}
+
+	picovoice.rhino = rhn.Rhino{
+		ModelPath:   picovoice.RhinoModelPath,
+		ContextPath: picovoice.ContextPath,
+	}
+	err = picovoice.rhino.Init()
+	if err != nil {
+		return err
+	}
+	picovoice.ContextInfo = picovoice.rhino.ContextInfo
+	picovoice.initialized = true
+	return nil
+}
+
+// Releases resouces aquired by Picovoice
+func (picovoice *Picovoice) Delete() error {
+	err := picovoice.porcupine.Delete()
+	if err != nil {
+		return err
+	}
+
+	err = picovoice.rhino.Delete()
+	if err != nil {
+		return err
+	}
+
+	picovoice.initialized = false
+	return nil
+}
+
+// Process a frame of pcm audio with the Picovoice platform.
+// Invokes user-defined callbacks upon detection of wake word and completion of follow-on command inference
+func (picovoice *Picovoice) Process(pcm []int16) error {
+	if !picovoice.initialized {
+		return fmt.Errorf("Picovoice could not process because it has either not been initialized or has been deleted.")
+	}
+
+	if len(pcm) != FrameLength {
+		return fmt.Errorf("Input data frame size (%d) does not match required size of %d", len(pcm), FrameLength)
+	}
+
+	if !picovoice.wakeWordDetected {
+		keywordIndex, err := picovoice.porcupine.Process(pcm)
+		if err != nil {
+			return err
+		}
+
+		if keywordIndex == 0 {
+			picovoice.wakeWordDetected = true
+			if picovoice.WakeWordCallback != nil {
```

Comment:
I was thinking this should be optional because devs might not always need to the callback so no need to ask them to write an empty function just 'cause

Reply:
Does it make sense to use it w/out callback? I.e. they can go and read the intent manually? I don't see a likely scenario where you want nothing to happen after rhino triggers. Can lead to very very confused users and difficult to solve help requests. Let's err on the side of that instead (or even provide a default callback that simply logs the result).